File: 3rd lab.c
/**********************************
* ПИ-221                          *
* Иванов Тимофей                  *
* Перегрузка операций             *
* 20.2.2023                       *
**********************************/

//Указал формат c т.к. почему-то неправильно выделяет текст в cs

using Microsoft.Win32;
using System;
using System.ComponentModel;

namespace lab {

  class Overload {
    public static void Plus(int FirstDigit, int SecondDigit) {
      Console.WriteLine($"{FirstDigit} + {SecondDigit} = {FirstDigit + SecondDigit}");
    }
    public static void Plus(int FirstDigit, int SecondDigit, int ThirdDigit) {
      Console.WriteLine($"{FirstDigit} + {SecondDigit} + {ThirdDigit} = {FirstDigit + SecondDigit + ThirdDigit}");
    }
    public static void Multiply(int FirstDigit, int SecondDigit) {
      Console.WriteLine($"{FirstDigit} * {SecondDigit} = {FirstDigit * SecondDigit}");
    }
    public static void Multiply(float FirstDigit, float SecondDigit) {
      float Result = FirstDigit * SecondDigit;
      Console.WriteLine($"{FirstDigit} * {SecondDigit} = {Result}");
    }
    public static void Compare(int FirstDigit, int SecondDigit) {
      if (FirstDigit > SecondDigit) {
        Console.WriteLine($"{FirstDigit} > {SecondDigit}");
      } else if (SecondDigit > FirstDigit) {
        Console.WriteLine($"{FirstDigit} < {SecondDigit}");
      } else {
        Console.WriteLine("Числа равны");
      }
    }
    public static void Compare(float FirstDigit, float SecondDigit) {
      if (FirstDigit > SecondDigit) {
        Console.WriteLine($"{FirstDigit} > {SecondDigit}");
      } else if (SecondDigit > FirstDigit) {
        Console.WriteLine($"{FirstDigit} < {SecondDigit}");
      } else {
        Console.WriteLine("Числа равны");
      }
    }
    public static void CompareWithEqual(int FirstDigit, int SecondDigit) {
      if (FirstDigit > SecondDigit) {
        Console.WriteLine($"{FirstDigit} >= {SecondDigit}");
      } else {
        Console.WriteLine($"{FirstDigit} <= {SecondDigit}");
      }
    }
    public static void CompareWithEqual(float FirstDigit, float SecondDigit) {
      if (FirstDigit > SecondDigit) {
        Console.WriteLine($"{FirstDigit} >= {SecondDigit}");
      } else {
        Console.WriteLine($"{FirstDigit} <= {SecondDigit}");
      }
    }
    public static void Equal(int FirstDigit, int SecondDigit) {
      if (FirstDigit == SecondDigit) {
        Console.WriteLine($"{FirstDigit} == {SecondDigit}");
      } else {
        Console.WriteLine($"{FirstDigit} != {SecondDigit}");
      }
    }
    public static void Equal(float FirstDigit, float SecondDigit) {
      if (FirstDigit == SecondDigit) {
        Console.WriteLine($"{FirstDigit} == {SecondDigit}");
      } else {
        Console.WriteLine($"{FirstDigit} != {SecondDigit}");
      }
    }
    public static string Transform(int Digit) {
      return Convert.ToString(Digit);
    }
    public static int Transform(string Digit) {
      return Convert.ToInt32(Digit);
    }
    public static bool Bool(bool FirstBool, bool SecondBool) {
      return FirstBool && SecondBool;
    }
    public static bool Bool(bool FirstBool, bool SecondBool, bool ThirdBool) {
      return FirstBool && SecondBool && ThirdBool;
    }
    public static int Determinant(SquareMatrix Matrix, int MatrixSize) {
      int Result = 0;

      for (int FirstDigit = 0; FirstDigit < MatrixSize; ++FirstDigit) {
        int WhatAdd = 1;
        for (int SecondDigit = 0; SecondDigit < MatrixSize; ++SecondDigit) {
          int SecondPos = FirstDigit + SecondDigit;
          if (SecondPos > MatrixSize - 1) SecondPos = SecondPos - MatrixSize;
          WhatAdd *= Matrix.Digits[SecondDigit, SecondPos];
        }
        Result += WhatAdd;
      }

      for (int FirstDigit = MatrixSize - 1; FirstDigit >= 0; --FirstDigit) {
        int WhatDelete = 1;
        for (int SecondDigit = 0; SecondDigit < MatrixSize; ++SecondDigit) {
          int SecondPos = FirstDigit - SecondDigit;
          if (SecondPos < 0) SecondPos = SecondPos + MatrixSize;
          WhatDelete *= Matrix.Digits[SecondDigit, SecondPos];
        }
        Result -= WhatDelete;
      }

      return Result;
    }

    public static int Determinant(SquareMatrix Matrix) {
      int Result = 0;

      int MatrixSize = Matrix.Size;
      for (int FirstDigit = 0; FirstDigit < MatrixSize; ++FirstDigit) {
        int WhatAdd = 1;
        for (int SecondDigit = 0; SecondDigit < MatrixSize; ++SecondDigit) {
          int SecondPos = FirstDigit + SecondDigit;
          if (SecondPos > MatrixSize - 1) SecondPos = SecondPos - MatrixSize;
          WhatAdd *= Matrix.Digits[SecondDigit, SecondPos];
        }
        Result += WhatAdd;
      }

      for (int FirstDigit = MatrixSize - 1; FirstDigit >= 0; --FirstDigit) {
        int WhatDelete = 1;
        for (int SecondDigit = 0; SecondDigit < MatrixSize; ++SecondDigit) {
          int SecondPos = FirstDigit - SecondDigit;
          if (SecondPos < 0) SecondPos = SecondPos + MatrixSize;
          WhatDelete *= Matrix.Digits[SecondDigit, SecondPos];
        }
        Result -= WhatDelete;
      }

      return Result;
    }

    public static void ReverseMatrix(SquareMatrix Matrix, int Size) {
      Console.WriteLine("Обратная матрица: ");
      Console.Write("\n1 / " + Overload.Determinant(Matrix) + " *\n");
      SquareMatrix NewMatrix = new SquareMatrix();
      for (int FirstDigit = 0; FirstDigit < Size; ++FirstDigit) {
        for (int SecondDigit = 0; SecondDigit < Size; ++SecondDigit) {
          int FirstFlag = 0;
          int SecondFlag = 0;
          for (int NewFirstDigit = 0; NewFirstDigit < Size - 1; ++NewFirstDigit) {
            for (int NewSecondDigit = 0; NewSecondDigit < Size - 1; ++NewSecondDigit) {         
              if (NewFirstDigit == FirstDigit) FirstFlag = 1;
              if (NewSecondDigit == SecondDigit)  SecondFlag = 1; 
              NewMatrix.Digits[NewFirstDigit, NewSecondDigit] = Matrix.Digits[NewFirstDigit + FirstFlag, NewSecondDigit + SecondFlag];
            }
          }
          int Result = Convert.ToInt32(Math.Pow(-1.0, Convert.ToDouble(FirstDigit * SecondDigit))) * Overload.Determinant(NewMatrix);
          Console.Write(Result + " ");
        }
        Console.Write("\n");
      }
    }
  }

  public class BigSize : System.Exception {
    public BigSize()
    : base() { }
    public BigSize(string message)
    : base(message) { }
    public BigSize
    (string message, System.Exception inner)
    : base(message, inner) { }

  }

  class SquareMatrix {
    public int Size;
    public int[,] Digits = new int[10, 10];
    public void ToString(SquareMatrix Matrix) {
      String Result = "";
      for (int FirstDigit = 0; FirstDigit < Matrix.Size; ++FirstDigit) {
        for (int SecondDigit = 0; SecondDigit < Matrix.Size; ++SecondDigit) {
          Result += Matrix.Digits[FirstDigit, SecondDigit] + " ";
        }
      Result += "\n";
      }
      Console.WriteLine(Result);
    }
    
    public int CompareTo(int FirstDigit, int SecondDigit) {
      if (FirstDigit > SecondDigit) return 1;
      if (FirstDigit < SecondDigit) return -1;
      return 0;
    }

    public SquareMatrix Prototip() => new SquareMatrix();

  }
  internal class Program {
    static void Main(string[] args) {
      Console.Write("введите размер матрицы(до 10):");
      SquareMatrix MyMatrix = new SquareMatrix();
      MyMatrix.Size = Convert.ToInt32(Console.ReadLine());
      if (MyMatrix.Size > 10) throw new BigSize("Слишком большой размер");
      Console.Clear();
      Random Random = new Random();
      for (int Row = 0; Row < MyMatrix.Size; ++Row) {
        for (int Column = 0; Column < MyMatrix.Size; ++Column) {
          MyMatrix.Digits[Row, Column] = Random.Next(0, 10);
          Console.Write(MyMatrix.Digits[Row, Column] + " ");
        }
        Console.Write("\n");
      }
      Console.Write("\n\n");

      Console.WriteLine("Перегрузка операций '+':");
      Overload.Plus(4, 5);
      Overload.Plus(4, 5, 6);
      Console.WriteLine("Перегрузка операции '*':");
      Overload.Multiply(2, 5);
      Overload.Multiply(2.4f, 7.2f);
      Console.WriteLine("Перегрузка операции '> и <':");
      Overload.Compare(2, 5);
      Overload.Compare(2.4f, 2.1f);
      Console.WriteLine("Перегрузка операции '>= и <=':");
      Overload.CompareWithEqual(7, 7);
      Overload.CompareWithEqual(45.1f, 5.2f);
      Console.WriteLine("Перегрузка операции '== и !='");
      Overload.Equal(2, 1);
      Overload.Equal(5.2f, 5.2f);
      Console.WriteLine("Перегрузка приведения типов:");
      Console.WriteLine("32 + 1 = " + (Overload.Transform(32) + 1));
      Console.WriteLine("32 + 1 = " + (Overload.Transform("32") + 1));
      Console.WriteLine("Перегрузка true false:");
      Console.WriteLine("True && False = " + Overload.Bool(true, false));
      Console.WriteLine("True && True && True = " + Overload.Bool(true, true, true));
      Console.WriteLine("Детерминант матрицы(с указанием размера) = " + Overload.Determinant(MyMatrix, MyMatrix.Size));
      Console.WriteLine("Детерминант матрицы(без указания размера) = " + Overload.Determinant(MyMatrix));
      MyMatrix.ToString(MyMatrix);
      Console.WriteLine("CompareTo: " + MyMatrix.CompareTo(MyMatrix.Digits[0, 0], MyMatrix.Digits[0, 1]));
      SquareMatrix MyMatrix2 = new SquareMatrix();
      for (int Row = 0; Row < MyMatrix.Size; ++Row) {
        for (int Column = 0; Column < MyMatrix.Size; ++Column) {
          MyMatrix.Digits[Row, Column] = Random.Next(0, 10);
        }
      }
      Console.WriteLine(MyMatrix.Equals(MyMatrix2));
      Console.WriteLine(MyMatrix.GetHashCode());
      SquareMatrix NewMyMatrix = MyMatrix.Prototip();
      
      Console.ReadKey();
      
    }
  }
}

File: 3rd lab.cs
/**********************************
* ПИ-221                          *
* Иванов Тимофей                  *
* ООП на C#                       *
* 13.2.2023                       *
**********************************/

using Microsoft.Win32;
using System;
using System.ComponentModel;
using System.Runtime.CompilerServices;
using System.Security.Cryptography.X509Certificates;

namespace lab {
  class Overload {
    public static int Determinant(SquareMatrix Matrix) {
      int Result = 0;
      for (int FirstDigit = 0; FirstDigit < Matrix.Size; ++FirstDigit) {
        int WhatAdd = 1;
        for (int SecondDigit = 0; SecondDigit < Matrix.Size; ++SecondDigit) {
          int SecondPos = FirstDigit + SecondDigit;
          if (SecondPos > Matrix.Size - 1) SecondPos = SecondPos - Matrix.Size;
          WhatAdd *= Matrix.Digits[SecondDigit, SecondPos];
        }
        Result += WhatAdd;
      }

      for (int FirstDigit = Matrix.Size - 1; FirstDigit >= 0; --FirstDigit) {
        int WhatDelete = 1;
        for (int SecondDigit = 0; SecondDigit < Matrix.Size; ++SecondDigit) {
          int SecondPos = FirstDigit - SecondDigit;
          if (SecondPos < 0) SecondPos = SecondPos + Matrix.Size;
          WhatDelete *= Matrix.Digits[SecondDigit, SecondPos];
        }
        Result -= WhatDelete;
      }
      return Result;
    }

    public static void ReverseMatrix(SquareMatrix Matrix) {
      Console.WriteLine("Обратная матрица: ");
      Console.Write("\n1 / " + Overload.Determinant(Matrix) + " *\n");
      SquareMatrix NewMatrix = new SquareMatrix();
      for (int FirstDigit = 0; FirstDigit < Matrix.Size; ++FirstDigit) {
        for (int SecondDigit = 0; SecondDigit < Matrix.Size; ++SecondDigit) {
          int FirstFlag = 0;
          int SecondFlag = 0;
          for (int NewFirstDigit = 0; NewFirstDigit < Matrix.Size - 1; ++NewFirstDigit) {
            for (int NewSecondDigit = 0; NewSecondDigit < Matrix.Size - 1; ++NewSecondDigit) {
              if (NewFirstDigit == FirstDigit) {
                FirstFlag = 1;
              }
              if (NewSecondDigit == SecondDigit) {
                SecondFlag = 1;
              }
              NewMatrix.Digits[NewFirstDigit, NewSecondDigit] = Matrix.Digits[NewFirstDigit + FirstFlag, NewSecondDigit + SecondFlag];
            }
          }
          int Result = Convert.ToInt32(Math.Pow(-1.0, Convert.ToDouble(FirstDigit * SecondDigit))) * Overload.Determinant(NewMatrix);
          Console.Write(Result + " ");
        }
        Console.Write("\n");
      }
    }
  }

  public class BigSize : System.Exception {
    public BigSize()
    : base() { }
    public BigSize(string message)
    : base(message) { }
    public BigSize
    (string message, System.Exception inner)
    : base(message, inner) { }

  }

  class SquareMatrix {
    public int Size;
    public int[,] Digits = new int[10, 10];
    public int CompareTo(SquareMatrix FirstMatrix, SquareMatrix SecondMatrix) {
      if (FirstMatrix.Sum(FirstMatrix) > SecondMatrix.Sum(SecondMatrix)) return 1;
      if (FirstMatrix.Sum(FirstMatrix) < SecondMatrix.Sum(SecondMatrix)) return -1;
      return 0;
    }
    public static SquareMatrix operator+ (SquareMatrix Matrix, int Digit) {
      for (int RowIndex = 0; RowIndex < Matrix.Size; ++RowIndex) {
        for (int ColumnIndex = 0; ColumnIndex < Matrix.Size; ++ColumnIndex) {
          Matrix.Digits[RowIndex, ColumnIndex] = Digit + Matrix.Digits[RowIndex,ColumnIndex];
        }
      }
      return Matrix;
    }
    public static SquareMatrix operator* (SquareMatrix Matrix, int Digit) {
      for (int RowIndex = 0; RowIndex < Matrix.Size; ++RowIndex) {
        for (int ColumnIndex = 0; ColumnIndex < Matrix.Size; ++ColumnIndex) {
          Matrix.Digits[RowIndex, ColumnIndex] = Digit * Matrix.Digits[RowIndex, ColumnIndex];
        }
      }
      return Matrix;
    }
    public int Sum(SquareMatrix Matrix) {
      int Sum = 0;
      for (int RowIndex = 0; RowIndex < Matrix.Size; ++RowIndex) {
        for (int ColumnIndex = 0; ColumnIndex < Matrix.Size; ++ColumnIndex) {
          Sum += Matrix.Digits[RowIndex, ColumnIndex];
        }
      }
      return Sum;
    }
    public static bool operator> (SquareMatrix MatrixLeft, SquareMatrix MatrixRight) {
      if (MatrixLeft.Sum(MatrixLeft) > MatrixRight.Sum(MatrixRight)) {
        return true;
      } else {
        return false;
      }
    }
    public static bool operator<(SquareMatrix MatrixLeft, SquareMatrix MatrixRight) {
      if (MatrixLeft.Sum(MatrixLeft) < MatrixRight.Sum(MatrixRight)) {
        return true;
      } else {
        return false;
      }
    }
    public static bool operator>=(SquareMatrix MatrixLeft, SquareMatrix MatrixRight) {
      if (MatrixLeft.Sum(MatrixLeft) >= MatrixRight.Sum(MatrixRight)) {
        return true;
      } else {
        return false;
      }
    }
    public static bool operator<=(SquareMatrix MatrixLeft, SquareMatrix MatrixRight) {
      if (MatrixLeft.Sum(MatrixLeft) <= MatrixRight.Sum(MatrixRight)) {
        return true;
      } else {
        return false;
      }
    }
    public void ToString(SquareMatrix Matrix) {
      for (int FirstDigit = 0; FirstDigit < Matrix.Size; ++FirstDigit) {
        for (int SecondDigit = 0; SecondDigit < Matrix.Size; ++SecondDigit) {
          Console.Write("{0,5}", Matrix.Digits[FirstDigit, SecondDigit]);
        }
        Console.WriteLine();
      }
    }
    public static bool operator==(SquareMatrix MatrixLeft, SquareMatrix MatrixRight) {
      if (MatrixLeft.Sum(MatrixLeft) == MatrixRight.Sum(MatrixRight)) {
        return true;
      } else {
        return false;
      }
    }
    public static bool operator!=(SquareMatrix MatrixLeft, SquareMatrix MatrixRight) {
      if (MatrixLeft.Sum(MatrixLeft) != MatrixRight.Sum(MatrixRight)) {
        return true;
      } else {
        return false;
      }
    }
    public static explicit operator string(SquareMatrix Matrix) {
      string Result = "";
      for (int FirstDigit = 0; FirstDigit < Matrix.Size; ++FirstDigit) {
        for (int SecondDigit = 0; SecondDigit < Matrix.Size; ++SecondDigit) {
          Result += "   " + Convert.ToString(Matrix.Digits[FirstDigit, SecondDigit]);
          Console.Write("{0,5}", Matrix.Digits[FirstDigit, SecondDigit]);
        }
        Result += "\n";
      }
      return Result;
    }
    public static bool operator true(SquareMatrix Matrix) {
      bool Result = true;
      for (int FirstDigit = 0; FirstDigit < Matrix.Size; ++FirstDigit) {
        for (int SecondDigit = 0; SecondDigit < Matrix.Size; ++SecondDigit) {
          if (Matrix.Digits[FirstDigit, SecondDigit] == 0) {
            Result = false;
          }
        }
      }
      return Result;
    }
    public static bool operator false(SquareMatrix Matrix) {
      bool Result = false;
      for (int FirstDigit = 0; FirstDigit < Matrix.Size; ++FirstDigit) {
        for (int SecondDigit = 0; SecondDigit < Matrix.Size; ++SecondDigit) {
          if (Matrix.Digits[FirstDigit, SecondDigit] == 0) {
            Result = true;
          }
        }
      }
      return Result;
    }
    public override bool Equals(object other) {
      bool result = false;
      if (other is SquareMatrix) {
        var param = other as SquareMatrix;
        if ((param.Digits == this.Digits) && (param.Size == this.Size))
          result = true;
      }
      return result;
    }
    public override int GetHashCode() {
      return (int)this.Digits[0,0];
    }
  }
  class Prototip {
    public int Size;
    public int[,] Digits = new int[10, 10];
    public SquareMatrix ds;
    public Prototip() {
      ds = new SquareMatrix();
    }
  }
  internal class Program {
    static void Main(string[] args) {
      Console.Write("введите размер матрицы(до 10):");
      SquareMatrix MyMatrix = new SquareMatrix();
      MyMatrix.Size = Convert.ToInt32(Console.ReadLine());
      if (MyMatrix.Size > 10) throw new BigSize("Слишком большой размер");
      Console.Clear();
      Random Random = new Random();
      SquareMatrix SecondMatrix = new SquareMatrix();
      for (int RowIndex = 0; RowIndex < MyMatrix.Size; ++RowIndex) {
        for (int ColumnIndex = 0; ColumnIndex < MyMatrix.Size; ++ColumnIndex) {
          MyMatrix.Digits[RowIndex, ColumnIndex] = Random.Next(0, 10);
          SecondMatrix.Digits[RowIndex, ColumnIndex] = Random.Next(0, 10);
        }
      }
      Console.WriteLine("Ваша матрица:");
      MyMatrix.ToString(MyMatrix);
      SecondMatrix.Size = MyMatrix.Size;

      Console.WriteLine("Выберите действие:\n1 - прибавить к матрице число\n2 - умножить матрицу на число\n3 - операция >\n4 - операция <\n5 - операция >=\n6 - операция <=\n7 - операция ==");
      Console.WriteLine("8 - операция !=\n9 - true/false\n10 - вывести детерминант");
      int Choice = Convert.ToInt32(Console.ReadLine());
      switch (Choice) {
        case 1:
          Console.WriteLine("Введите число, которое нужно прибавить: ");
          int DigitPlus = Convert.ToInt32(Console.ReadLine());
          MyMatrix.ToString(MyMatrix + DigitPlus);
          break;
        case 2:
          Console.WriteLine("Введите число, на которое нужно умножить: ");
          int DigitMultiply = Convert.ToInt32(Console.ReadLine());
          MyMatrix.ToString(MyMatrix * DigitMultiply);
          break;
        case 3:
          Console.WriteLine("Вторая матрица: ");
          SecondMatrix.ToString(SecondMatrix);
          Console.WriteLine(MyMatrix > SecondMatrix);
          break;
        case 4:
          SecondMatrix.ToString(SecondMatrix);
          Console.WriteLine(MyMatrix < SecondMatrix);
          break;
        case 5:
          SecondMatrix.ToString(SecondMatrix);
          Console.WriteLine(MyMatrix >= SecondMatrix);
          break;
        case 6:
          SecondMatrix.ToString(SecondMatrix);
          Console.WriteLine(MyMatrix <= SecondMatrix);
          break;
        case 7:
          SecondMatrix.ToString(SecondMatrix);
          Console.WriteLine(MyMatrix == SecondMatrix);
          break;
        case 8:
          SecondMatrix.ToString(SecondMatrix);
          Console.WriteLine(MyMatrix != SecondMatrix);
          break;
        case 9:
          if (MyMatrix) {
            Console.WriteLine("в матрице нету нулей");
          } else {
            Console.WriteLine("в матрице есть нули");
          }
          break;
        case 10:
          Console.WriteLine(Overload.Determinant(MyMatrix));
          break;
      }
      Console.ReadKey();
    }
  }
}
